### ISSUE TITLE
Connection#open() promise rejection is broken

### ISSUE THREAD

**Issue by daltones** [Automattic/mongoose#5229]
**Do you want to request a *feature* or report a *bug*?**
Report a bug

**What is the current behavior?**
When `Connection#open()` fails, it is emitting an error event that turns into an uncaught exception, breaking the promise rejection handling.

**If the current behavior is a bug, please provide the steps to reproduce.**
```js
import mongoose from 'mongoose';

const conn = mongoose.createConnection(); // don't open now

async function start() {
  try {
    await conn.open('some wrong connection string');
    // this promise will throw
    // but before that an error event is emitted, leading to an uncaught exception
  } catch (error) {
    console.log('error on open conn');
  }
}

start();
```

I'm guessing the problem is here: https://github.com/Automattic/mongoose/blob/2dd76ab803b809ab7be2b6881256a6527fd3e11c/lib/connection.js#L316  
That `promise.$hasHandler` is not valid here, once `$hasHandler` is a property of `MongooseThenable` instances (which is not the case there).

This affects `Connection#openSet()` too.

**What is the expected behavior?**
While open returns a promise, I expect that any error appears only as the rejection reason and not thrown somewhere else.

**Please mention your node.js, mongoose and MongoDB version.**
node.js: v7.10.0
mongoose: v4.9.7
MongoDB: v3.4.3

**Comment by vkarpov15** [comment]
As a workaround, add an `on('error')` handler. 

```javascript
const mongoose = require('mongoose');

const conn = mongoose.createConnection(); // don't open now

conn.on('error', () => {}); // Add an error handler to prevent the exception

async function start() {
  try {
    await conn.open('some wrong connection string');
    // this promise will throw
    // but before that an error event is emitted, leading to an uncaught exception
  } catch (error) {
    console.log('error on open conn');
  }
}

start();
```

In general, I recommend not using `open` and `openSet` directly unless you're certain you want to connect to a standalone. Use `createConnection()` with a URI param instead because that'll figure out replica set connections etc.

```javascript
const mongoose = require('mongoose');

async function start() {
  try {
    const conn = await mongoose.createConnection('some wrong connection string');
    // this promise will throw
    // but before that an error event is emitted, leading to an uncaught exception
  } catch (error) {
    console.log('error on open conn');
  }
}

start();
```

**Comment by vkarpov15** [comment]
Fix will be in 4.9.9. Monkey-patching `.then()` hurts but is necessary
